pgx/src/pgx.ml

```diff
@@ -526,7 +531,59 @@ module Make (Thread : Io) = struct
 
   (*----- Connection. -----*)
 
+  let attempt_tls_upgrade ?(ssl = `Auto) ({ ichan; chan; _ } as conn) =
+    (* To initiate an SSL-encrypted connection, the frontend initially sends an SSLRequest message rather than a
+       StartupMessage. The server then responds with a single byte containing S or N, indicating that it is willing
+       or unwilling to perform SSL, respectively. The frontend might close the connection at this point if it is
+        dissatisfied with the response. To continue after S, perform an SSL startup handshake (not described here,
+       part of the SSL specification) with the server. If this is successful, continue with sending the usual
+       StartupMessage. In this case the StartupMessage and all subsequent data will be SSL-encrypted. To continue
+       after N, send the usual StartupMessage and proceed without encryption.
+       See https://www.postgresql.org/docs/9.3/protocol-flow.html#AEN100021 *)
+    match ssl with
+    | `No -> return conn
+    | (`Auto | `Always _) as ssl ->
+      (match Io.upgrade_ssl with
+      | `Not_supported ->
+        (match ssl with
+        | `Always _ ->
+          failwith
+            "TLS support is not compiled into this Pgx library but ~ssl was set to \
+             `Always"
+        | _ -> ());
+        debug
+          "TLS-support is not compiled into this Pgx library, not attempting to upgrade"
+        >>| fun () -> conn
+      | `Supported upgrade_ssl ->
+        debug "Request SSL upgrade from server"
+        >>= fun () ->
+        let msg = Message_out.SSLRequest in
+        send_message conn msg
+        >>= fun () ->
+        flush chan
+        >>= fun () ->
+        input_char ichan
+        >>= (function
+        | 'S' ->
+          debug "Server supports TLS, attempting to upgrade"
+          >>= fun () ->
+          let ssl_config =
+            match ssl with
+            | `Auto -> None
+            | `Always ssl_config -> Some ssl_config
```

Comment:
I'm not completely sure on this interace. My idea was that if people actually care about SSL, they probably want to send a config.